Batch automation of a ARIAC trials + user system

dhood opened the pull request:
Addresses https://github.com/osrf/ariac-docker/issues/3

This adds the [run_trial.bash script](https://github.com/osrf/ariac-docker/blob/235fdbfdce5c17d508b3fc70960f573581133ca4/run_trial.bash) that can be used to bringup the server and competitor containers and run a given trial config file. It also shuts the two containers down automatically, provided that the competitors system gets the competition into a 'done' state (by completing the trial, making a call to the `/ariac/end_competition` service (giving up), or time running out).

There's also the start of a script to call `run_trial.bash` with the different trial names but something's not right with it at the moment.

dhood committed:
WIP batch mode

dhood committed:
Add script from caguero

dhood committed:
Install killall

dhood committed:
Mount config and log dirs

dhood committed:
Use ARIAC_EXIT_ON_COMPLETION instead of wait_until_competition_ends

dhood committed:
Run competitor container too

dhood committed:
Re-order container bringup so that when the competition dies it brings down the user's container

dhood committed:
Parameterise scripts and give example in readme

dhood committed:
Add timeout to example trials so server is sure to die

dhood committed:
Add script to run all trials (doesn't pass args correctly ATM)

dhood commented on ariac-server/ariac-server/run_ariac_task.sh:
```diff
@@ -20,20 +20,6 @@ usage()
   exit 1
 }
 
-# Wait until the /ariac/competition_state returns "data: done".
```
@caguero I modified the script from https://bitbucket.org/osrf/ariac/pull-requests/58 to use the `ARIAC_EXIT_ON_COMPLETION` var instead of checking for 'done'. This should bring the simulation down when the trial is complete, time's up, or the user calls `/ariac/end_competition`

dhood committed:
Add call to ariac_network.bash

dhood committed:
Don't fetch example submission from gist

dhood committed:
Add TODOs

dhood commented on run_trial.bash:
```diff
@@ -0,0 +1,47 @@
+# Make the log playable
+#TODO: figure out why the ownership of the state log is root and resolve.
+echo "Changing ownership of gazebo state log"
+sudo chown $USER ${HOST_LOG_DIR}/gazebo/state.log
```
@j-rivero, I noticed that the gazebo state log has `root` ownership. The log is generated in the container, and then copied into `/ariac/logs` on the container [in this script](https://github.com/osrf/ariac-docker/pull/6/files#diff-861a436c3d2af41ee884b625a77cd85aR47), which is [mounted to `logs/{team_name}/{trial_name}` on the host](https://github.com/osrf/ariac-docker/pull/6/files#diff-c2e963c2b9be2d58754e5afa23675caeR38). Do you have any idea what might cause this?

This workaround is necessary to make the log playable on the host. If we can avoid it, the script won't need `sudo` I believe.

j-rivero replied:
are we running gazebo as root in the container? AFAIK there is no other change in permissions of the state log inside gazebo.

dhood replied:
Gazebo gets launched from a command passed through [this line in the entrypoint](https://github.com/osrf/ariac-docker/pull/6/commits/aa6899f2992350e1c4ced7c270d25e4788d0548d#diff-6b2aa3d92ed12b835c3dcc8a9d6429afR19). I see now that it looks to be run as root user: https://github.com/osrf/ariac-docker/blob/aa6899f2992350e1c4ced7c270d25e4788d0548d/ariac-server/ariac-server/Dockerfile#L2. Maybe changing that will fix it (still seems strange to me that it wouldn't have the permissions of the mounted volume owner, but I haven't worked with mounting much before).

dhood replied:
I changed the the entrypoint to be run as the ariac user and now it's all good! f011bac4ab6e05fc5ed4722f10a76347d35a22f0

dhood replied:
oops, got lucky on my machine with userid. Added https://github.com/osrf/ariac-docker/pull/6/commits/2c3a8705c291e6d99f59f7d86027a1db4c67e9bd for portability (thanks @dirk-thomas)

dhood committed:
[untested] Disable nvidia support

dhood commented:
@caguero: I added the example code back into the repository. I added some TODOs in https://github.com/osrf/ariac-docker/pull/6/commits/aa6899f2992350e1c4ced7c270d25e4788d0548d. I think we spoke about most of them in person.

The most important one is https://github.com/osrf/ariac-docker/pull/6/commits/aa6899f2992350e1c4ced7c270d25e4788d0548d#diff-92694b972d13716f9b231735929abe9bR6 (the example code has to sleep for 10s at the moment because the competitor container is started first). We still don't support kinetic at the moment, but we don't know how many users are using kinetic to know how big of a priority it should be.

I added https://github.com/osrf/ariac-docker/pull/6/commits/8ed5087f241e8b8f0324dfaec8391d5316b91dae which should disable the nvidia support but I can't test it because the airport internet won't let me pull docker images. Please revert it if is doesn't work!

dhood commented:
@j-rivero would you be able to give this a look if you get a chance please? There are some known to-do's that I've noted but you might be able to spot some other issues since you have more experience with Docker than @caguero and I.

I forgot to document that if you want to stop the competition manually as opposed to waiting for it to time out (10mins by default!) you can run `docker exec -it ariac-competitior-system /bin/bash` to get into​ the user's container, and then source the ROS setup file and call the `/ariac/end_competition` service. It should then kill the simulation and bring down the competition container as a result.

caguero committed:
Add instructions for installing Docker.

caguero committed:
Remove version from ros-indigo-ros-core and ros-indigo-ros-base while building gazebo Docker image.

caguero committed:
Update README

caguero committed:
Update README.

caguero commented:
I had to remove the version number while installing `ros-indigo-ros-core` and `ros-indigo-ros-base` packages (building a Docker image). This is probably due to a new ROS sync. I also added a note about installing `Docker`.

j-rivero commented on ariac-competitor/build_competitor_image.bash:
```diff
@@ -2,8 +2,22 @@
 
 DOCKER_ARGS=""
 # Uncoment this line to rebuild without cache
+# TODO: expose this as an argument
```
what do we want to expose as an argument here?

dhood replied:
it doesn't show up well in the diffs, sorry: [the no-cache option](https://github.com/osrf/ariac-docker/pull/6/files/1d787ab18d87b9530456c46ac696ee2fc90d3fc2#diff-a3996c3db93ff76671e5a51b86508a1cR6)

j-rivero replied:
I'm using the `no-cache` option in my upcoming `kinetic_version` branch, let's handle it from there.

dhood replied:
ok, we'll just have to do it for the competitor script as well as the server script

j-rivero commented on run_all_trials.bash:
```diff
@@ -0,0 +1,11 @@
+
+TEAM_NAME=$1
+
+#TODO: get list of trials automatically
```
where is the list of trails or how we can get it?

dhood replied:
From `ls comp_configs` minus the extensions. each of the config files in the comp_configs directory is used for a different trial (once).

j-rivero replied:
Oh, it was pretty obvious sorry for the question. Let's try a patch to fix all the three comments/problems we have in this file:

```diff
diff --git a/run_all_trials.bash b/run_all_trials.bash
index 3f0822e..7fd7b7c 100755
--- a/run_all_trials.bash
+++ b/run_all_trials.bash
@@ -1,10 +1,24 @@
 #!/usr/bin/env bash
 
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
 TEAM_NAME=$1
+COMP_CONFIGS_DIR=${DIR}/comp_configs/
+
+get_list_of_trials()
+{
+  yaml_files=$(ls ${COMP_CONFIGS_DIR}/*.yaml)
+
+  for f in $(ls ${COMP_CONFIGS_DIR}/*.yaml); do
+    f=${f##*/}
+    f=${f//.yaml}
+    all_names="${all_names} ${f}"
+  done
+
+  echo $all_names
+}
 
-#TODO: get list of trials automatically
-declare -a TRIAL_NAMES=("example_trial1" "example_trial2")
-for TRIAL_NAME in ${TRIAL_NAMES[@]}; do
+for TRIAL_NAME in $(get_list_of_trials); do
   echo "Running trial: ${TRIAL_NAME}"
   #TODO: GET THIS TO RUN THE INDIVIDUAL TRIALS CORRECTLY.
   ./run_trial.bash ${TEAM_NAME} ${TRIAL_NAME}
```

dhood replied:
thanks, I'll try this!

dhood replied:
added in aa62c7e81a2bf7fb4868861b203767b005f3040e thanks

j-rivero commented on run_all_trials.bash:
```diff
@@ -0,0 +1,11 @@
+TEAM_NAME=$1
+
+#TODO: get list of trials automatically
+declare -a TRIAL_NAMES=("example_trial1" "example_trial2")
```
we probably don't need a bash array here, should be enough using:
```bash
TRIAL_NAMES="example_trial1 example_trial2"
 for TRIAL_NAME in ${TRIAL_NAMES}; do
...
```

dhood replied:
thank you for your advice here (and elsewhere!) -- indeed that looks like a simpler way to achieve it

j-rivero commented on run_all_trials.bash:
```diff
@@ -0,0 +1,11 @@
+declare -a TRIAL_NAMES=("example_trial1" "example_trial2")
+for TRIAL_NAME in ${TRIAL_NAMES[@]}; do
+  echo "Running trial: ${TRIAL_NAME}"
+  #TODO: GET THIS TO RUN THE INDIVIDUAL TRIALS CORRECTLY.
```
What is the problem here?

dhood replied:
when the `run_trial.bash` script gets called from here, inside the `run_trial.bash` script the `TRIAL_NAME` variable sometimes has the correct value, but sometimes is empty. (whereas if you call the script from the terminal, the variable is always correct). I didn't have heaps of time to investigate it thoroughly yet - maybe your suggestion to not use bash arrays is enough to fix it.

j-rivero replied:
let's try to the patch I've posted above and see if it is enough or we need to debug deeper.

dhood replied:
in the end quoting the parameters fixed it: 2b1a20b52e46207af5b43314a1453d08ce56ef64

j-rivero commented on run_trial.bash:
```diff
@@ -0,0 +1,47 @@
+
+# Start the competitors container and let it run in the background.
+#TODO: parameterize the container name
+./ariac-competitor/run_competitor_container.bash &
```
This background call to run_competitor_container script is going to be hard to check and response on errors. For example I forget to generate the team container and run the script. It keeps going even if this call fails totally which ends up in other errors more difficult to debug. We can workaround with some checks before.

dhood replied:
thank you for the explanation of why this could be an issue, I wasn't aware

j-rivero replied:
I don't have a good suggestion to improve the situation right now, but let's keep this in mind.

dhood replied:
I have encountered a different problem because of this: if the server doesn't start correctly (next line), the script exits, and the competitor container is left running in the background. I added a script for users to kill the containers in 019b15c664de2a96d1dee34f73e2d3f045bfd583 which will at least let people handle the situation

j-rivero commented on run_trial.bash:
```diff
@@ -0,0 +1,47 @@
+
+# Start the competitors container and let it run in the background.
+#TODO: parameterize the container name
+./ariac-competitor/run_competitor_container.bash &
```
Will make more sense to launch first the ariac server and the competitor code after it?

dhood replied:
This part of the script is where I am most looking for feedback. It definitely would be preferable to have the competitor code start after the ariac server, so that we don't need to do [waiting like this](https://github.com/osrf/ariac-docker/pull/6/commits/aa6899f2992350e1c4ced7c270d25e4788d0548d#diff-92694b972d13716f9b231735929abe9bR6) in the competitor code.

The reason the competitor code is launched first is for the flow of the script. I launch the competitor code in the background first, and it may run indefinitely. The server, however, will only run for the duration of the trial (limited by a timeout): I launch it in the foreground, it is blocking, and when it ends, we know the trial is finished, so we kill the competitor container.

I am "utilising" the blocking nature of the server here, but it is not ideal, it was just the easiest way I knew how to do it! If you could help me launch the competitor code after the server (and then kill it once the trial has ended), I would like to learn from how you do it.

j-rivero replied:
Your approach makes sense although I think we are abusing a little bit of the nature of the system instead of creating a proper sync mechanism. A possible solution could be to implement a service in the ariac server that triggers the beginning of the trial. This service could be called from the user code (as soon as the team consider the code is ready) or by a timeout by our system (let's say give X minutes to startup).

It is hard to automatically try to know when are all the pieces ready using ROS1.

dhood replied:
Yeah, it's a bit complicated: the team's do have to call a service etc, but I believe this issue has more to do with the ROS master not being available at the time. If the team's container starts before the server (which is the ROS master), when the ARIAC simulation starts it gives the following error:

```
Loaded ariac/arm
Started ['ariac/arm'] successfully
Traceback (most recent call last):
  File "/opt/ros/indigo/lib/controller_manager/controller_manager", line 63, in <module>
    controller_manager_interface.load_controller(c)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/controller_manager/controller_manager_interface.py", line 66, in load_controller
    resp = s.call(LoadControllerRequest(name))
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 525, in call
    raise ServiceException("transport error completing service call: %s"%(str(e)))
rospy.service.ServiceException: transport error completing service call: unable to receive data from sender, check sender's logs for details
```

I'll see if moving the ROS master around helps...

dhood replied:
For the record, I tried moving it to the competitor container and to a third container but it still gives that error when the server connects to the master. I left the server container as the ROS master and moved the waiting to a workaround script in https://github.com/osrf/ariac-docker/pull/8/commits/64b10ed22d8048c2b6b950906b3bbfd34c2721ee

j-rivero commented:
Most of my comments are doubts or minor things. I would like to submit some changes and the implementation of the nvidia-docker detection. Feel free to merge and I will submit them against master or I can create the PR against this branch, whatever you prefer.

Important things still to do (in my list):

 * Implement the nvidia-docker detection and use of nvidia-docker mechanism
 * Implement the mechanism for the hardware acceleration in [docker using Intel graphics cards](http://wiki.ros.org/docker/Tutorials/Hardware%20Acceleration)
 * Implement the option of having kinetic

j-rivero commented:
update:

> Implement the nvidia-docker detection and use of nvidia-docker mechanism
> Implement the mechanism for the hardware acceleration in docker using Intel graphics cards

these two should not be hard to implement.

> Implement the option of having kinetic

see https://github.com/osrf/ariac-docker/issues/2#issuecomment-303140511

dhood committed:
Get the list of available trials automatically

dhood committed:
Exit on errors

dhood committed:
Quote arguments

dhood committed:
Document run_all_trials

dhood committed:
Add script for killing ARIAC containers

dhood committed:
Don't pass display params unnecessarily

Avoids the following error when starting gazebo on my machine:

libGL error: failed to load driver: swrast
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  154 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  30
  Current serial number in output stream:  31

dhood committed:
Print docker run args

dhood committed:
Capturing of stdout and stderr can be done by a higher-level script

dhood committed:
Don't execute ariac entrypoint as root

dhood commented:
Thanks @j-rivero for your help with this PR and #2. I've managed to address most issues now, the main one left is https://github.com/osrf/ariac-docker/pull/6#discussion_r117572375: I'll make a follow-up PR.

> Implement the nvidia-docker detection and use of nvidia-docker mechanism
> Implement the mechanism for the hardware acceleration in docker using Intel graphics cards

For now we have disabled the use of GPU completely in this PR so that we can get it out the door, since users can still replay the logs post-hoc. It's just considered a "nice-to-have": we're happy to drop support for it if it's complicated to implement. No need for intel support.

> Implement the option of having kinetic
> see #2 (comment)

Awesome! testing now

dhood committed:
Link to non-root docker instructions

dhood committed:
Ensure that files written to mounted directories have the same userid

Otherwise copying ARIAC logs to the mounted directory is not guaranteed to have correct permissions

dhood committed:
Show docker build parameters
